sway-lib-std/src/storage.sw

```diff
@@ -638,3 +639,91 @@ impl<V> StorageVec<V> {
         store(__get_storage_key(), 0);
     }
 }
+
+pub struct StorageBytes {}
+
+impl StorageBytes {
+    /// Takes a `Bytes` type and stores the underlying collection of tightly packed bytes.
+    ///
+    /// ### Arguments
+    ///
+    /// * `bytes` - The bytes which will be stored.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     stored_bytes: StorageBytes = StorageBytes {}
+    /// }
+    ///
+    /// fn foo() {
+    ///     let mut bytes = Bytes::new();
+    ///     bytes.push(5_u8);
+    ///     bytes.push(7_u8);
+    ///     bytes.push(9_u8);
+    ///
+    ///     storage.stored_bytes.write_bytes(bytes);
+    /// }
+    /// ```
+    #[storage(write)]
+    pub fn write_bytes(self, mut bytes: Bytes) {
```

Comment:
Is there a reason to not use `write()` and instead to specifically outline that this function is used for storing bytes?

Reply:
Not one that I feel strongly about. I flip-flopped between `from_bytes` and `store_bytes` and `write_bytes`. Considering we have an `into_bytes` it may make more sense to use `from_bytes`.

Would like to hear feedback from others as well as to what is most appropriate here.

Reply:
There are already `write` & `write_byte` methods on the `raw_ptr` type... 
Perhaps going with `store_bytes` here to make it really clear. Unless we want to follow what `StorageVec` has done with `push`, `pop`, etc. Would these operations be more expensive for `StorageBytes` than they are for `StorageVec` ?

Reply:
I've specifically avoided adding `push` operations here but have renamed the function to `store_bytes` for clarity.

The `push()` operations are inherently more expensive than `StorageVec`'s push, especially for large collections of bytes. Take the following as an example. Here we are pushing 3 numbers to a `StorageVec`. This involves 3 read/write calls, where each call is writing a new element with a key(the length) and a value. This involves 3 separate storage slots for each element pushed. Importantly, the previous elements are not read to the heap.

```rust
storage {
     vec: StorageVec<u64> = StorageVec {}
}

fn foo() {
     storage.vec.push(5);
     storage.vec.push(7);
     storage.vec.push(9);
}
```

`StorageBytes` does not store with variable keys(length) so we are using the same storage slot. In the following example, every `push()` would involve loading the existing bytes to the heap, appending a byte, and writing the new bytes into the same storage slot.

```rust
storage {
     stored_bytes: StorageBytes = StorageBytes {}
}

fn foo() {
     storage.stored_bytes.push(5_u8);
     storage.stored_bytes.push(7_u8);
     storage.stored_bytes.push(9_u8);
}
```

Instead, this can be done entirely in the heap and will require less gas, especially for longer bytes.

```rust
storage {
     stored_bytes: StorageBytes = StorageBytes {}
}

fn foo() {
     let mut bytes = Bytes::new();
     bytes.push(5_u8);
     bytes.push(7_u8);
     bytes.push(9_u8);

     storage.stored_bytes.store_bytes(bytes);
}
```